### PR TITLE
internal/{v4, legacy}: serve metadata.yaml with empty series

### PR DESCRIPTION
**Pull request by rogpeppe**
Multi-series charms contain a list of supported series in the
"series" field in the metadata.yaml file. This field has changed
type, so some juju versions prior to 1.26 cannot parse the string
list form. To ensure compatibility with old version without
changing the name of the field, in legacy and v4 API versions,
we elide the series field from multi-series charms (for those
API versions, the series is always available in the charm id).

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1505/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1506/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1507/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1508/
Test PASSed.

**Comment by mhilton** [comment]
:+1:

File: internal/charmstore/archive.go [hunk 1]
```diff
@@ -4,16 +4,229 @@
...
+		Hash:           hash,
+	}, nil
+}
+
+type multiReadSeekCloser struct {
+	readers []blobstore.ReadSeekCloser
+	io.ReadSeeker
+}
+
+func newMultiReadSeekCloser(readers ...blobstore.ReadSeekCloser) blobstore.ReadSeekCloser {
```
Comment (frankban) [inline internal/charmstore/archive.go:91]: Nice enough to deserve a docstring?

File: internal/charmstore/hash.go [hunk 2]
```diff
@@ -1,53 +0,0 @@
-// Copyright 2015 Canonical Ltd.
-// Licensed under the AGPLv3, see LICENCE file for details.
-
-// TODO frankban: remove this file after updating entities in the production db
```
Comment (frankban) [inline internal/charmstore/hash.go:4]: Thanks for deleting this.

File: internal/charmstore/archive.go (hunk 1 above)
Comment (rogpeppe) [inline internal/charmstore/archive.go:91]: I _think_ it should be obvious enough from the signature and the similarity to io.MultiReader.

File: internal/charmstore/store.go [hunk 3]
```diff
@@ -1073,22 +986,31 @@ func (s StoreDatabase) Collections() []*mgo.Collection {
 	return cs
 }
 
+// readerAtSeeker adapts an io.ReadSeeker to an io.ReaderAt.
 type readerAtSeeker struct {
-	r io.ReadSeeker
+	r   io.ReadSeeker
+	off int64
```
Comment (frankban) [inline internal/charmstore/store.go:992]: So this is to avoid unnecessary Seek calls?

File: internal/charmstore/store_test.go [hunk 4]
```diff
@@ -334,31 +334,36 @@ func (s *StoreSuite) TestFindEntities(c *gc.C) {
...
+	c.Assert(entity0.Size, gc.Not(gc.Equals), 0)
+
+	// Check that it doesn't matter if it's a development URL.
+	rurl.Development = true
+	entity1, err := store.FindEntity(rurl, nil)
+	c.Assert(err, gc.IsNil)
+	c.Assert(entity1, jc.DeepEquals, entity0)
+
+	// Check that the field selector works.
+
```
Comment (frankban) [inline internal/charmstore/store_test.go]: Unrequired newline.

File: internal/charmstore/addentity.go [hunk 5]
```diff
@@ -169,12 +189,124 @@ func (s *Store) addEntityFromReader(id *router.ResolvedURL, r io.ReadSeeker, blo
...
+		return errgo.Notef(err, "cannot open metadata.yaml from archive")
+	}
+	defer fr.Close()
+	data, err := removeSeriesField(fr)
+	if err != nil {
+		return errgo.Notef(err, "cannot remove series field from metadata")
+	}
+	var appendedBlob bytes.Buffer
+	zw := z.Append(&appendedBlob)
+	updatedf := metadataf.FileHeader // Work around invalid duplicate FileHeader issue
```
Comment (frankban) [inline internal/charmstore/addentity.go]: s/issue/issue./

File: internal/mongodoc/doc.go [hunk 6]
```diff
@@ -60,6 +76,9 @@ type Entity struct {
 	Size int64
 
 	// BlobName holds the name that the archive blob is given in the blob store.
+	// For multi-series charms, there is also a second blob which
+	// stores a "zip-suffix" that overrides metadata.yaml.
+	// This is named BlobName + ".pre-v5-suffix".
```
Comment (frankban) [inline internal/mongodoc/doc.go:81]: This is not true, it seems to me that it is named BlobName + ".suffix". And perhaps it should change to what we describe here.

File: internal/v4/archive.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
+package v4
```
Comment (frankban) [inline internal/v4/archive.go]: Missing license header.

File: internal/v5/archive.go [hunk 8]
```diff
@@ -304,52 +300,38 @@ func (h *ReqHandler) latestRevisionInfo(id *charm.URL) (*router.ResolvedURL, str
...
-		w.Header().Set("Content-Length", strconv.FormatInt(fileInfo.Size(), 10))
-		setArchiveCacheControl(w.Header(), h.isPublic(id.URL))
-		w.WriteHeader(http.StatusOK)
-		io.Copy(w, content)
-		return nil
+// ServeBlobFile serves a file from the given blob. The
+// path of the file is taken from req.URL.Path.
+// The blob should be associated with the entity
+// with the given id.
+//
```
Comment (frankban) [inline internal/v5/archive.go]: Leftover?

File: internal/v4/api.go [hunk 9]
```diff
@@ -225,9 +240,33 @@ func (h *ReqHandler) metaRevisionInfo(id *router.ResolvedURL, path string, flags
 	return &response, nil
 }
 
+// GET id/meta/archive-size
+// https://github.com/juju/charmstore/blob/v4/docs/API.md#get-idmetaarchive-size
+func (h ReqHandler) metaArchiveSize(entity *mongodoc.Entity, id *router.ResolvedURL, path string, flags url.Values, req *http.Request) (interface{}, error) {
+	return &params.ArchiveSizeResponse{
+		Size: entity.PreV5BlobSize,
```
Comment (frankban) [inline internal/v4/api.go:247]: This is nice, but it seems we have no migrations to populate those fields.

**Comment by frankban** [comment]
:+1: looks very nice.
I have some minor comments and a more important concern about missing migrations.
Thank you!

File: internal/charmstore/store.go (hunk 3 above)
Comment (rogpeppe) [inline internal/charmstore/store.go:992]: yes

File: internal/v4/api.go (hunk 9 above)
Comment (rogpeppe) [inline internal/v4/api.go:247]: migrations coming in the next PR.

File: internal/charmstore/store_test.go (hunk 4 above)
Comment (rogpeppe) [inline internal/charmstore/store_test.go]: done

File: internal/charmstore/addentity.go (hunk 5 above)
Comment (rogpeppe) [inline internal/charmstore/addentity.go]: done

File: internal/mongodoc/doc.go (hunk 6 above)
Comment (rogpeppe) [inline internal/mongodoc/doc.go:81]: done

File: internal/v4/archive.go (hunk 7 above)
Comment (rogpeppe) [inline internal/v4/archive.go]: done

File: internal/v5/archive.go (hunk 8 above)
Comment (rogpeppe) [inline internal/v5/archive.go]: no, this is still used.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1509/
Test PASSed.

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**